File: Medium/1870. Minimum Speed to Arrive on Time/Solution.java
class Solution {
    final int oo = 10000005;

    boolean can(int[] dist, double hour, int mid) {
        int n = dist.length;
        double ret = 0;

        for(int i = 0; i < n - 1; i++) {
            double val = (double) dist[i] / (double) mid;
            val = Math.ceil(val);
            ret += val;
        }
        ret += (double) dist[n - 1] / (double) mid;

        return ret <= hour;
    }

    public int minSpeedOnTime(int[] dist, double hour) {
        int ans = oo;
        int l = 1, r = oo, mid;

        while(l <= r) {
            mid = (l + r) / 2;
            if(can(dist, hour, mid)) {
                ans = mid;
                r = mid - 1;
            } else l = mid + 1;
        }

        return (ans == oo ? -1 : ans);
    }
}

File: Medium/1870. Minimum Speed to Arrive on Time/Solution.py
class Solution(object):

    def can(self, dist, hour, mid):
        ret = 0.0
        for i in range(len(dist) - 1):
            ret += math.ceil(dist[i] / mid)
        ret += dist[-1] / mid
        return ret <= hour

    def minSpeedOnTime(self, dist, hour):
        oo = 10**7 + 5
        l, r, ans = 1, oo, oo
        while l <= r:
            mid = (l + r) // 2
            if self.can(dist, hour, mid):
                ans = mid
                r = mid - 1
            else:
                l = mid + 1
        return ans if ans < oo else -1